pkg/storage/formattedstore/batchwrite.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package formattedstore
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+var valueIsDeletedErr = fmt.Errorf("value is deleted")
+
+// BatchService is a batch service.
+type BatchService struct {
+	sizeLimit int
+	timeLimit time.Duration
+	*sync.RWMutex
+	values       []*batch
+	futureValues map[string]*batch
+	formatter    Formatter
+}
+
+type batch struct {
+	keyID     string
+	value     []byte
+	isDeleted bool
+}
+
+// NewBatchWrite new batch write
+func NewBatchWrite(sizeLimit int, timeLimit time.Duration, formatter Formatter) *BatchService {
+	ticker := time.NewTicker(timeLimit)
+	quit := make(chan struct{})
+	b := &BatchService{sizeLimit: sizeLimit, timeLimit: timeLimit, values: make([]*batch, 0),
+		futureValues: make(map[string]*batch), formatter: formatter}
+	go func(b *BatchService) {
+		for {
+			select {
+			case <-ticker.C:
+				if err := b.flush(); err != nil {
+					logger.Errorf(err.Error())
+				}
+			case <-quit:
+				ticker.Stop()
+				return
+			}
+		}
+	}(b)
+
+	return b
+}
+
+// Get value from batch values
+func (b *BatchService) Get(k string) ([]byte, error) {
+	b.RLock()
+	v, ok := b.futureValues[k]
+	b.RUnlock()
+	if !ok {
+		return nil, fmt.Errorf("k %s not found", k)
+	}
+
+	if v.isDeleted {
+		return nil, valueIsDeletedErr
+	}
+
+	return v.value, nil
+}
+
+// Put in batch value
+func (b *BatchService) Put(k string, v []byte) error {
+	b.RLock()
+	size := len(b.values)
+	b.RUnlock()
+
+	if size >= b.sizeLimit {
+		if err := b.flush(); err != nil {
+			return err
+		}
+	}
+
+	b.Lock()
+	b.futureValues[k] = &batch{keyID: k, value: v, isDeleted: false}
+	b.Unlock()
+
+	go func() {
+		formattedValue, err := b.formatter.FormatPair(k, v) // 300ms
+		if err != nil {
+			logger.Errorf("failed to format pair for k=%s: %w", k, err)
+			return
+		}
+
+		b.Lock()
+		b.values = append(b.values, &batch{keyID: k, value: formattedValue, isDeleted: false})
+		b.Unlock()
+	}()
+
+	return nil
+}
+
+func (b *BatchService) Delete(k string) error {
+	b.RLock()
+	size := len(b.values)
+	b.RUnlock()
+
+	if size >= b.sizeLimit {
+		if err := b.flush(); err != nil {
```

Comment:
There's no delete? I don't understand why this check is done here.

Reply:
I will remove it

Reply:
done